Foundation/NSNotification.swift

```diff
@@ -57,32 +58,43 @@ private let _defaultCenter: NSNotificationCenter = NSNotificationCenter()
 
 public class NSNotificationCenter : NSObject {
     
-    private var observers: [NSNotificationReceiver]
-    
-    public required override init() {
-        observers = [NSNotificationReceiver]()
-    }
+    private var _observers = [NSNotificationReceiver]()
+    private let _observersLock = NSLock()
     
     public class func defaultCenter() -> NSNotificationCenter {
         return _defaultCenter
     }
     
     public func postNotification(notification: NSNotification) {
+
         let name = notification.name
         let sender = notification.object
-        
-        let sendTo = observers.filter { observer in
+
+        _observersLock.lock()
+        let sendTo = _observers.filter { observer in
             let sameName = (observer.name == nil || observer.name == name)
             let sameSender = (observer.sender == nil || observer.sender === sender)
             
             return sameSender && sameName
         }
+        _observersLock.unlock()
         
         for observer in sendTo {
             guard let block = observer.block else {
                 continue
             }
             
+            if let queue = observer.queue {
+                
+                let operation = NSBlockOperation {
+                    block(notification)
+                }
+                
+                queue.addOperation(operation)
+
+                return
```

Comment:
Also we cannot return in here, otherwise we will only notify the first observer registered with a queue when they may be multiple or in another example if the first observer has a queue but the second one doesn't.